src/modules-load/modules-load.c

```diff
@@ -52,12 +73,31 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
         return 0;
 }
 
-static int apply_file(struct kmod_ctx *ctx, const char *path, bool ignore_enoent) {
+static void exec_task(int *tasks, int *r, thread_data *data, char *line, const bool to_free) {
+        int i = *tasks;
+        while (*tasks >= max_tasks) {
+                for (i = 0; i < max_tasks; ++i) {
+                        if (pthread_tryjoin_np(data[i].thread, NULL) == 0) {
```

Comment:
if this fails because there's no thread that already finised you'll be busy looping here? that sounds quite wrong?

again, please distribute work properly among your worker threads/processes, I'd recommend via a AF_UNIX/SOCK_SEQPACKET socketpair(), because it's really easy to do: just create the socket, then send() on it from the parent process, once for each module. And in the workers just do recv() on them until you see EOF.

Reply:
This is implemented in the latest commit and is faster than all other solutions.